tests/integration/plugins/test_ml_plugins.py

```diff
@@ -0,0 +1,256 @@
+                    "dimension": 4,
+                    "method": {
+                        "name": "hnsw",
+                        "space_type": "l2",
+                        "engine": "nmslib",
+                        "parameters": {"ef_construction": 256, "m": 48},
+                    },
+                }
+            }
+        },
+    )
+    payload, payload_list = generate_bulk_training_data(
+        100, 4, index_name, vector_name, has_result=True
+    )
+    # Insert data in bulk
+    await bulk_insert(ops_test, app, leader_unit_ip, payload)
+    query = {"size": 2, "query": {"knn": {vector_name: {"vector": payload_list[0], "k": 2}}}}
+    s = await search(ops_test, app, leader_unit_ip, index_name, query)
+    assert len(s) == 2
+    assert "query_shard_exception" not in json.dumps(s)
+
+
+@pytest.mark.abort_on_fail
+async def test_train_search_with_ivf_faiss(ops_test: OpsTest) -> None:
+    """Uploads data and runs a query search against the FAISS KNNEngine."""
+    app = (await app_name(ops_test)) or APP_NAME
+
+    units = await get_application_unit_ids_ips(ops_test, app=app)
+    leader_unit_ip = await get_leader_unit_ip(ops_test, app=app)
+
+    # create index with r_shards = nodes - 1
+    index_name = "test_train_search_with_ivf_faiss_training"
+    vector_name = "test_train_search_with_ivf_faiss_vector"
+    model_name = "test_train_search_with_ivf_faiss_model"
+    await create_index_and_bulk_insert(
+        ops_test, app, leader_unit_ip, index_name, len(units) - 1, vector_name
+    )
+
+    # train the model
+    await run_knn_training(
+        ops_test,
+        app,
+        leader_unit_ip,
+        "test_train_search_with_ivf_faiss_model",
+        {
+            "training_index": index_name,
+            "training_field": vector_name,
+            "dimension": 4,
+            "method": {
+                "name": "ivf",
+                "engine": "faiss",
+                "space_type": "l2",
+                "parameters": {"nlist": 4, "nprobes": 2},
+            },
+        },
+    )
+
+    # wait for training to finish
+    finished = False
+    for i in range(10):
+        if await wait_for_knn_training(ops_test, app, leader_unit_ip, model_name):
+            finished = True
+            break
+        print("test_train_search_with_ivf_faiss - Waiting for training to finish...")
+        await asyncio.sleep(15)
+
+    assert finished
+
+
+@pytest.mark.abort_on_fail
+async def test_disable_re_enable_knn(ops_test: OpsTest) -> None:
+    """Disables the KNN plugin, check restart happened, test unreachable and re-enable it."""
+    app = (await app_name(ops_test)) or APP_NAME
+
+    units = await get_application_unit_ids_ips(ops_test, app=app)
+    leader_unit_ip = await get_leader_unit_ip(ops_test, app=app)
+    # Get since when each unit has been active
+    active_since = {unit: None for unit in ops_test.model.applications[APP_NAME].units}
+    active_since = await wait_all_units_restarted_or_fail(active_since)
```

Comment:
Could we slightly change  and breakdown the logic to: 
1. getting the start time of the daemon by unit. (i.e `units_start_time = {unit_id: start_time(unit_id) for unit_id in get_application_units(ops_test)}`)
2. trigger the restart
3. repeat # 1
4. compare both dictionaries
5. unsure all the nodes are up `cluster_formation_successful`

Reply:
Changed it to get one single timestamp, then all the config change happens and compare each opensearch.